torchtune/utils/_generation.py

```diff
@@ -4,141 +4,115 @@
-        )
+    prompt = prompt.view(1, -1) if prompt.ndim == 1 else prompt
+    stop_tokens = (
+        torch.tensor(stop_tokens, device=prompt.device) if stop_tokens else None
+    )
+    bsz, seq_length = prompt.size()
+    generated_tokens = prompt.clone()
+    stop_token_reached = torch.zeros(bsz, dtype=torch.bool, device=prompt.device)
+    stop_token_mask = torch.ones(
+        (bsz, seq_length + 1), dtype=torch.int32, device=prompt.device
+    )
 
     if custom_generate_next_token is None:
         custom_generate_next_token = generate_next_token
 
-    # generated_tokens is a list of tensors where each tensor contains tokens
-    # needed for the output
-    generated_tokens = [prompt]
-
-    # generate the first token by conditioning on the input prompt
+    # generate the first token conditioned on the prompt
     token = generate_next_token(
-        model=model,
-        input_pos=torch.arange(0, prompt_length, device=prompt.device),
-        # convert the input into [B, S] shape as expected by the model
-        x=prompt.view(1, -1),
+        model,
+        input_pos=torch.arange(0, seq_length, device=prompt.device),
+        x=prompt,
         temperature=temperature,
         top_k=top_k,
-    ).clone()
+    )
+    generated_tokens = torch.cat([generated_tokens, token], dim=-1)
 
-    generated_tokens.append(token)
+    # stop early if we reach a stop token in every seq
+    if stop_tokens is not None:
+        stop_token_reached_curr = torch.isin(token, stop_tokens)
+        stop_token_reached |= stop_token_reached_curr.flatten()
+        if stop_token_reached.all().item():
+            return generated_tokens.tolist()
 
-    # generation starts at position=prompt_length and continues till
-    # we get the requested number of tokens or we hit eos_id
-    input_pos = torch.tensor([prompt_length], device=prompt.device)
+    input_pos = torch.tensor([seq_length], device=prompt.device)
     for _ in range(max_generated_tokens - 1):
+        # update stop_token_mask if we reached a stop token in a previous step
+        if stop_tokens is not None:
+            stop_token_mask = torch.cat(
+                [stop_token_mask, ~stop_token_reached.reshape(bsz, 1)], dim=-1
+            )
+
         token = custom_generate_next_token(
-            model=model,
-            input_pos=input_pos,
-            x=token.view(1, -1),
-            temperature=temperature,
-            top_k=top_k,
-        ).clone()
+            model, input_pos=input_pos, x=token, temperature=temperature, top_k=top_k
+        )
 
-        if stop_tokens is not None and token.item() in stop_tokens:
-            break
+        generated_tokens = torch.cat([generated_tokens, token], dim=-1)
+        input_pos += 1
 
-        generated_tokens.append(token)
+        if stop_tokens is not None:
+            stop_token_reached_curr = torch.isin(token, stop_tokens)
+            stop_token_reached |= stop_token_reached_curr.flatten()
```

Comment:
nit: just put flatten in the preceding line, otherwise the ior is a bit more confusing